client/src/mid.rs

```diff
@@ -302,92 +678,60 @@ impl State {
     }
 
     /* pub fn register_event(&mut self, name: &str) {
-        todo!()
+        Default::default()
     }
 
     pub fn event_notify(&mut self, name: &str) -> bool {
-        todo!()
+        Default::default()
     } */
 }
 
 // request helper function
 #[tracing::instrument]
-async fn do_request<Req, Res>(req_builder: RequestBuilder, req: Req) -> color_eyre::Result<Res>
+async fn do_request<Req, Res>(
+    req_builder: RequestBuilder,
+    req: Req,
+) -> reqwest_middleware::Result<Res>
 where
     Req: Serialize + std::fmt::Debug,
     Res: for<'d> Deserialize<'d> + std::fmt::Debug,
 {
     let res: Response = req_builder.json(&req).send().await?;
-    let bytes = res.bytes().await?.to_vec();
-    let res: Res = serde_json::from_reader(&bytes[..]).with_context(|| {
-        format!(
-            "should have received type {}, as json, received: \"{}\"",
-            std::any::type_name::<Res>(),
-            String::from_utf8_lossy(&bytes)
-        )
-    })?;
+    let bytes = res.bytes().await?;
+    tracing::debug!("received data: {bytes:?}");
+    let res: Res = serde_json::from_slice(&bytes).map_err(reqwest_middleware::Error::middleware)?;
     Ok(res)
 }
 
 /// Init middleware state
 /// This function is called by UI to create the Middleware state and establish a connection to the Database.
 /// Important: Make sure `url` does not contain a trailing `/`
 #[tracing::instrument]
-pub async fn init(url: &str) -> color_eyre::Result<State> {
-    let mut state = State {
-        url: url.to_owned(),
-        ..Default::default()
-    };
+pub fn init(url: &str) -> color_eyre::Result<(State, Receiver<MidEvent>)> {
```

Comment:
THis was changed so that the receiver gets returned with the state and simplified right? this is a very critical part of code here can i get a quick summary of this change?